nucypher/network/server.py

```diff
@@ -334,73 +336,92 @@ def reencrypt_via_rest(id_as_hex):
 
-        treasure_map_index = bytes.fromhex(treasure_map_id)
-
-        # First let's see if we already have this map.
-
-        try:
-            previously_saved_map = this_node.treasure_maps[treasure_map_index]
-        except KeyError:
-            pass # We don't have the map.  We'll validate and perhaps save it.
-        else:
-            if previously_saved_map == treasure_map:
-                return Response("Already have this map.", status=303)
-                # Otherwise, if it's a different map with the same ID, we move on to validation.
-
-        if treasure_map.public_id() == treasure_map_id:
-            do_store = True
+        # Additionally, we determine the map identifier from the type of node. 
+        # If the node is federated, we also set the expiration for a week.
+        if not this_node.federated_only:
+            map_identifier = received_treasure_map._hrac.hex()
         else:
-            return Response("Can't save a TreasureMap with this ID from you.", status=409)
+            map_identifier = received_treasure_map.public_id()
+            expiration_date = MayaDT.from_datetime(datetime.utcnow() + timedelta(days=7))
 
-        if do_store and not this_node.federated_only:
-            alice_checksum_address = this_node.policy_agent.contract.functions.getPolicyOwner(
-                treasure_map._hrac[:16]).call()
-            do_store = treasure_map.verify_blockchain_signature(checksum_address=alice_checksum_address)
+        # Step 2: Check if we already have the treasure map.
+        try:
+            with datastore.describe(TreasureMap, map_identifier) as stored_treasure_map:
+                if _MapClass.from_bytes(stored_treasure_map.treasure_map) == received_treasure_map:
+                    return Response("Already have this map.", status=303)
+        except RecordNotFound:
+            # This appears to be a new treasure map that we don't have!
+            pass
 
-        if do_store:
-            log.info("{} storing TreasureMap {}".format(this_node, treasure_map_id))
-            this_node.treasure_maps[treasure_map_index] = treasure_map
-            return Response(bytes(treasure_map), status=202)
-        else:
-            log.info("Bad TreasureMap ID; not storing {}".format(treasure_map_id))
-            return Response("This TreasureMap doesn't match a paid Policy.", status=402)
+        # Step 3: If the node is decentralized, we check that the received
+        # treasure map is valid pursuant to an active policy.
+        # We also set the expiration from the data on the blockchain here.
+        if not this_node.federated_only:
+            policy_data, alice_checksum_address = this_node.policy_agent.fetch_policy(
+                    received_treasure_map._hrac,
+                    with_owner=True)
+            # If the Policy doesn't exist, the policy_data is all zeros.
+            if not policy_data[5]:
```

Comment:
Can this be a reference to `NULL_ADDRESS` or another constant to be more explicit?

Reply:
I'm not really sure how that would apply here -- the data it returns is literally an `int` of `0`.

Reply:
> the policy_data is all zeros

Oh I see - I was thrown off by this comment.  This can be more readable with a named tuple type but that's better left for a future PR IMO.